src/EventStore.Core/TransactionLog/Chunks/TFChunk/TFChunk.cs

```diff
@@ -369,20 +360,8 @@ private enum CacheStatus {
 		// We therefore only read from memory while the chunk is still being written to, and only create
 		// the file streams when the chunk is being completed.
 		private void CreateReaderStreams() {
-			Interlocked.Add(ref _fileStreamCount, _internalStreamsCount);
-
-			if (_selfdestructin54321)
-				throw new FileBeingDeletedException();
-
-			for (int i = 0; i < _internalStreamsCount; i++) {
-				_fileStreams.Add(CreateInternalReaderWorkItem());
-			}
-		}
-
-		private ReaderWorkItem CreateInternalReaderWorkItem() {
-			Debug.Assert(_handle is not null);
-
-			return new(_handle);
+			_fileStreams = new();
+			Interlocked.Add(ref _fileStreamCount, _fileStreams.Count);
```

Comment:
here also, i'm not 100% sure but it appears that we may still want to do the increment operation before creating the pool as it seems possible for `CreateReaderStreams()` to execute concurrently with `TryDestructFileStreams()`

we may consider making the size a constant in `ReaderWorkItemPool` so that we know it in advance

Reply:
> we may consider making the size a constant in ReaderWorkItemPool so that we know it in advance

It's a subject for the potential bug in the future. `IndexPool` can be instantiated with any capacity that is less than or equal to `IndexPool.Capacity` property. To change the capacity, we need to change one line of code. Having constant, we need to keep it synchronized. Moreover, `LOCK.XADD` with constant operand doesn't work faster.

Reply:
```csharp
_fileStreams = new();
Interlocked.Add(ref _fileStreamCount, _fileStreams.Count);
```

To make this code a safe when concurrency with `TryDestructFileStreams` occurs, I need to insert `if` statement to check `localReferenceCount` on equality with 0. This check creates a correct happens-before relationship. The same cannot be done if switch those two lines of code. As a result, I prefer to keep the order.

Reply:
im satisfied with everything now except for the threading implications of the two Drain calls, which might be ok but i've run out of steam to tell for today

i'm slightly suspicious because the Drain() can be called on _fileStreams and _memStreams when they are not initialised (this is ok i think) and concurrently with them being initialized, which im less sure about, especially with respect to getting the counting right. it might be worth a few more words to explain why its safe

Reply:
`Drain` itself is thread safe, because all instance methods of `IndexPool` are thread safe. The first who reaches `Take` will take all available indices and dispose them. The pool is empty after `Take(Span<int>)`.

Reply:
The only thing that is not thread safe is initialization of `ReaderWorkItemPool`.

Reply:
Following the change in 45e4a3029629d152c5345627fd61ebd314c74a05 (the > 0 check), is an interleaving like this possible where `_fileStreams.Drain()` could return `0` causing `CleanUpFileStreamDestruction();` to be called but there might still be active filestreams?

```
1. CreateReaderStreams() :: _fileStreams = new(); // active chunk completed, reader streams being created
2. GetReaderWorkItem() (internally calls _fileStreams.TryTake) // triggered by read
3. TryDestructFileStreams() // triggered by manual shutdown, _fileStreams.Drain() will return 0.
4. CreateReaderStreams() :: Interlocked.Add(ref _fileStreamCount, _fileStreams.Count);
```

i'm not sure if step 2 would be able to create a file stream due to `IsReadOnly` not being set to true yet (i assume that the barriers in Interlocked.Add and _fileStreams.TryTake could prevent that from happening)

> To change the capacity, we need to change one line of code. Having constant, we need to keep it synchronized.

alternatively, we could also add a `capacity` constructor parameter to ReaderWorkItemPool to avoid the above